instrumentor/controllers/sourceinstrumentation/common.go

```diff
@@ -0,0 +1,329 @@
+	// migration code, add this condition to previous instrumentation configs
+	// which were created before this condition was introduced
+	if len(ic.Status.RuntimeDetailsByContainer) > 0 {
+		ic.Status.Conditions = append(ic.Status.Conditions, metav1.Condition{
+			Type:               v1alpha1.RuntimeDetectionStatusConditionType,
+			Status:             metav1.ConditionTrue,
+			Reason:             string(v1alpha1.RuntimeDetectionReasonWaitingForDetection),
+			Message:            "runtime detection completed successfully",
+			LastTransitionTime: metav1.NewTime(time.Now()),
+		})
+		return true
+	}
+
+	// if the workload has no available replicas, we can't detect the runtime
+	if workloadObj.AvailableReplicas() == 0 {
+		ic.Status.Conditions = append(ic.Status.Conditions, metav1.Condition{
+			Type:               v1alpha1.RuntimeDetectionStatusConditionType,
+			Status:             metav1.ConditionFalse,
+			Reason:             string(v1alpha1.RuntimeDetectionReasonNoRunningPods),
+			Message:            "No running pods available to detect source runtime",
+			LastTransitionTime: metav1.NewTime(time.Now()),
+		})
+		return true
+	}
+
+	ic.Status.Conditions = append(ic.Status.Conditions, metav1.Condition{
+		Type:               v1alpha1.RuntimeDetectionStatusConditionType,
+		Status:             metav1.ConditionUnknown,
+		Reason:             string(v1alpha1.RuntimeDetectionReasonWaitingForDetection),
+		Message:            "Waiting for odiglet to initiate runtime detection in a node with running pod",
+		LastTransitionTime: metav1.NewTime(time.Now()),
+	})
+
+	return true
+}
+
+func initiateAgentEnabledConditionIfMissing(ic *v1alpha1.InstrumentationConfig) bool {
+	if meta.FindStatusCondition(ic.Status.Conditions, v1alpha1.AgentEnabledStatusConditionType) != nil {
+		// avoid adding the condition if it already exists
+		return false
+	}
+
+	ic.Status.Conditions = append(ic.Status.Conditions, metav1.Condition{
+		Type:               v1alpha1.AgentEnabledStatusConditionType,
+		Status:             metav1.ConditionUnknown,
+		Reason:             string(v1alpha1.AgentEnabledReasonWaitingForRuntimeInspection),
+		Message:            "Waiting for runtime detection to complete",
+		LastTransitionTime: metav1.NewTime(time.Now()),
+	})
+
+	return true
+}
+
+// giving the input conditions array, this function will return a new array with the conditions sorted by logical order
+func sortIcConditionsByLogicalOrder(conditions []metav1.Condition) []metav1.Condition {
+	slices.SortFunc(conditions, func(i, j metav1.Condition) int {
+		return v1alpha1.StatusConditionTypeLogicalOrder(i.Type) - v1alpha1.StatusConditionTypeLogicalOrder(j.Type)
+	})
+	return conditions
+}
+
+func deleteWorkloadInstrumentationConfig(ctx context.Context, kubeClient client.Client, workloadObject client.Object) error {
+	logger := log.FromContext(ctx)
+	ns := workloadObject.GetNamespace()
+	name := workloadObject.GetName()
+	kind := workload.WorkloadKindFromClientObject(workloadObject)
```

Comment:
nit: the caller of `deleteWorkloadInstrumentationConfig` already knows the kind.

we can just pass the kind to this function and make it a bit simpler, saving the complexity of translating it back and forth. 

We actually don't even need the `client.Object`, only the ns+kind+name, and can consider just sending the `k8sconsts.PodWorkload`